File: src/kalshi_bot/edge_math.py
"""Fee-adjusted edge and intra-market boxed (YES+NO) surplus — Kalshi-only.

Interpreting **YES as shares** (see ``trading_model``): ``net_edge_buy_yes_long`` is expected edge
per **share** at the ask (dollars on the \$1 face); multiply by share count for total edge dollars
before fees on a sized order.
"""

from __future__ import annotations

from kalshi_bot.fees import kalshi_general_taker_fee_usd


def implied_yes_ask_dollars(best_no_bid_dollars: float) -> float:
    """Lift YES: ~ 1 − best NO bid (binary complement on the bid book)."""
    return max(0.01, min(0.99, 1.0 - _clamp01(best_no_bid_dollars)))


def implied_no_ask_dollars(best_yes_bid_dollars: float) -> float:
    """Lift NO: ~ 1 − best YES bid."""
    return max(0.01, min(0.99, 1.0 - _clamp01(best_yes_bid_dollars)))


def _clamp01(x: float) -> float:
    return max(0.0, min(1.0, x))


def boxed_pair_cost_dollars(best_yes_bid_dollars: float, best_no_bid_dollars: float) -> float:
    """Cost to buy 1 YES @ implied YES ask + 1 NO @ implied NO ask (before fees), in dollars."""
    return implied_yes_ask_dollars(best_no_bid_dollars) + implied_no_ask_dollars(best_yes_bid_dollars)


def boxed_arb_surplus_before_fees_dollars(best_yes_bid_dollars: float, best_no_bid_dollars: float) -> float:
    """1.0 − cost; positive means naive boxed arb before fees & rounding."""
    return 1.0 - boxed_pair_cost_dollars(best_yes_bid_dollars, best_no_bid_dollars)


def boxed_arb_surplus_after_taker_fees_dollars(
    best_yes_bid_dollars: float,
    best_no_bid_dollars: float,
    *,
    contracts: int = 1,
) -> float:
    """Surplus after paying taker fee on each leg (same contract count per leg)."""
    ya = implied_yes_ask_dollars(best_no_bid_dollars)
    na = implied_no_ask_dollars(best_yes_bid_dollars)
    pay = ya + na
    fy = kalshi_general_taker_fee_usd(contracts=contracts, price_dollars=ya)
    fn = kalshi_general_taker_fee_usd(contracts=contracts, price_dollars=na)
    return 1.0 - pay - (fy + fn) / max(1, contracts)


def net_edge_buy_yes_long(
    *,
    fair_yes: float,
    yes_ask_dollars: float,
    contracts: int = 1,
) -> float:
    """fair − ask − taker fee per share (all in dollars on \$1 face). ``contracts`` = share count for fee averaging."""
    fy = kalshi_general_taker_fee_usd(contracts=contracts, price_dollars=yes_ask_dollars)
    per = fy / max(1, contracts)
    return fair_yes - yes_ask_dollars - per


def middle_penalty_multiplier(mid: float, *, width: float = 0.15) -> float:
    """Extra edge (dollars) suggested near 0.50 — fees worst at mid; width = distance from 0.5 to start penalty."""
    d = abs(mid - 0.5)
    if d >= width:
        return 0.0
    return width - d


def min_edge_threshold_for_mid(
    mid: float,
    *,
    base_min_edge: float,
    middle_extra: float,
    middle_width: float = 0.15,
) -> float:
    """Require larger edge near 50% (fee + adverse selection heuristic)."""
    return base_min_edge + middle_extra * (middle_penalty_multiplier(mid, width=middle_width) / max(middle_width, 1e-9))

File: azure-wrapper/Bot/EdgeMath.cs
namespace KalshiBotWrapper.Bot;

/// <summary>
/// Fee-adjusted edge and intra-market boxed (YES+NO) surplus. Mirrors Python edge_math.py.
/// </summary>
public static class EdgeMath
{
    private static double Clamp01(double x) => Math.Max(0.0, Math.Min(1.0, x));

    public static double ImpliedYesAskDollars(double bestNoBidDollars)
        => Math.Max(0.01, Math.Min(0.99, 1.0 - Clamp01(bestNoBidDollars)));

    public static double ImpliedNoAskDollars(double bestYesBidDollars)
        => Math.Max(0.01, Math.Min(0.99, 1.0 - Clamp01(bestYesBidDollars)));

    public static double BoxedPairCostDollars(double bestYesBidDollars, double bestNoBidDollars)
        => ImpliedYesAskDollars(bestNoBidDollars) + ImpliedNoAskDollars(bestYesBidDollars);

    public static double BoxedArbSurplusBeforeFeesdDollars(double bestYesBidDollars, double bestNoBidDollars)
        => 1.0 - BoxedPairCostDollars(bestYesBidDollars, bestNoBidDollars);

    public static double BoxedArbSurplusAfterTakerFeesDollars(
        double bestYesBidDollars, double bestNoBidDollars, int contracts = 1)
    {
        var ya = ImpliedYesAskDollars(bestNoBidDollars);
        var na = ImpliedNoAskDollars(bestYesBidDollars);
        var pay = ya + na;
        var fy = Fees.KalshiGeneralTakerFeeUsd(contracts, ya);
        var fn = Fees.KalshiGeneralTakerFeeUsd(contracts, na);
        return 1.0 - pay - (fy + fn) / Math.Max(1, contracts);
    }

    /// <summary>fair − ask − taker fee per contract (all in dollars). Conservative long YES signal.</summary>
    public static double NetEdgeBuyYesLong(double fairYes, double yesAskDollars, int contracts = 1)
    {
        var fy = Fees.KalshiGeneralTakerFeeUsd(contracts, yesAskDollars);
        var per = fy / Math.Max(1, contracts);
        return fairYes - yesAskDollars - per;
    }

    /// <summary>Extra edge required near 0.50 (fees worst at mid).</summary>
    public static double MiddlePenaltyMultiplier(double mid, double width = 0.15)
    {
        var d = Math.Abs(mid - 0.5);
        if (d >= width) return 0.0;
        return width - d;
    }

    /// <summary>Require larger edge near 50% (fee + adverse selection heuristic).</summary>
    public static double MinEdgeThresholdForMid(
        double mid,
        double baseMinEdge,
        double middleExtra,
        double middleWidth = 0.15)
    {
        return baseMinEdge + middleExtra *
            (MiddlePenaltyMultiplier(mid, middleWidth) / Math.Max(middleWidth, 1e-9));
    }
}
